File: lib/openwook/services/practice.ts
import 'server-only';

import { sql } from '../db';
import { ApiError } from '../api';
import {
  acquireRedisLock,
  redisGetJson,
  redisGetOrSetJson,
  redisKey,
  redisSetJson
} from '../redis';
import { env } from '../env';
import {
  answerKeyCacheKey,
  fillBlankValues,
  getExistingPracticeAnswer,
  hasNonEmptyText,
  invalidateBankAnalytics,
  invalidateBankLeaderboard,
  invalidatePracticeSummaryCaches,
  invalidateUserAnalytics,
  loadBankQuestionItems,
  loadPracticeQuestionIds,
  normalizePracticeModeForTest,
  normalizeQuestionCountForTest,
  normalizeStringArray,
  objectValue,
  parseJson,
  practiceProgressFullLimit,
  practiceProgressWindowRadius,
  practiceQuestionQueueKey,
  practiceQueueTtl,
  toJsonValue
} from './internal';
import { getBank } from './banks';
import type { AnswerMode, PracticeAnswer, PracticeMode, PracticeSession, User } from '../types';

export async function startPracticeSession(
  user: User,
  data: {
    bankId: number;
    sessionType?: 'practice' | 'review' | 'exam';
    questionCount?: number;
    mode?: PracticeMode;
    questionTypeId?: string | null;
    allQuestions?: boolean;
  }
) {
  await getBank(user, data.bankId);
  const mode = normalizePracticeModeForTest(data.mode, data.sessionType);
  if (mode === 'by_type' && !data.questionTypeId?.trim()) {
    throw new ApiError(422, 'VALIDATION_ERROR', 'Question type is required for type-based practice');
  }
  const count = normalizeQuestionCountForTest(data.questionCount, data.allQuestions || mode === 'all' && data.questionCount === undefined);
  const questionIds = await loadPracticeQuestionIds(user, data.bankId, count, {
    mode,
    questionTypeId: data.questionTypeId ?? null,
    allQuestions: data.allQuestions
  });
  if (questionIds.length < 1) {
    throw new ApiError(
      409,
      'INVALID_STATE',
      mode === 'wrong'
        ? 'No wrong questions are available for review'
        : 'No active questions are available for practice'
    );
  }
  const rows = await sql<PracticeSession[]>`
    INSERT INTO user_practice_sessions (user_id, bank_id, session_type, question_count)
    VALUES (${user.id}, ${data.bankId}, ${mode === 'exam' ? 'exam' : mode === 'wrong' ? 'review' : data.sessionType ?? 'practice'}, ${questionIds.length})
    RETURNING *
  `;
  await redisSetJson(practiceQuestionQueueKey(rows[0].id), questionIds, practiceQueueTtl);
  return rows[0];
}

export async function getPracticeSession(user: User, sessionId: number) {
  const rows = await sql<PracticeSession[]>`
    SELECT *
    FROM user_practice_sessions
    WHERE id = ${sessionId} AND user_id = ${user.id}
    LIMIT 1
  `;
  if (!rows[0]) throw new ApiError(404, 'NOT_FOUND', 'Practice session not found');
  return rows[0];
}

async function ensurePracticeQuestionQueue(session: PracticeSession) {
  if (!session.bank_id) return [];
  const key = practiceQuestionQueueKey(session.id);
  const cachedQuestionIds = await redisGetJson<number[]>(key);
  if (cachedQuestionIds?.length) return cachedQuestionIds.map(Number).slice(0, session.question_count);

  const rows = await sql<Array<{ question_id: number }>>`
    WITH active_question_ids AS (
      SELECT bql.question_id, bql.sort_order AS bank_sort_order, NULL::integer AS group_sort_order
      FROM bank_question_links bql
      WHERE bql.bank_id = ${session.bank_id}
        AND bql.status = 'active'

      UNION ALL

      SELECT gql.question_id, bgl.sort_order AS bank_sort_order, gql.sort_order AS group_sort_order
      FROM bank_group_links bgl
      JOIN group_question_links gql ON gql.group_id = bgl.group_id
      WHERE bgl.bank_id = ${session.bank_id}
        AND bgl.status = 'active'
    )
    SELECT ids.question_id
    FROM active_question_ids ids
    JOIN questions q ON q.id = ids.question_id
    WHERE q.status = 'active'
    ORDER BY ids.bank_sort_order, ids.group_sort_order NULLS FIRST, ids.question_id
    LIMIT ${session.question_count}
  `;
  const questionIds = rows.map((row) => Number(row.question_id));
  await redisSetJson(key, questionIds, practiceQueueTtl);
  return questionIds;
}

async function loadPracticeQuestionRows(session: PracticeSession, questionIds: number[], offset = 0, limit = session.question_count) {
  if (!session.bank_id || questionIds.length === 0) return [];
  const slice = questionIds.slice(offset, offset + limit);
  if (slice.length === 0) return [];
  const orderedItems = await loadBankQuestionItems(session.bank_id, {
    questionIds: slice,
    activeOnly: true,
    limit: slice.length
  });
  const orderByQuestionId = new Map(slice.map((questionId, index) => [Number(questionId), index]));
  return orderedItems.sort(
    (left, right) => (orderByQuestionId.get(Number(left.question_id)) ?? 0) - (orderByQuestionId.get(Number(right.question_id)) ?? 0)
  );
}

export async function getPracticeQuestions(user: User, sessionId: number) {
  const session = await getPracticeSession(user, sessionId);
  const questionIds = await ensurePracticeQuestionQueue(session);
  return loadPracticeQuestionRows(session, questionIds);
}

export async function getPracticeQuestionPage(user: User, sessionId: number, params?: URLSearchParams) {
  const session = await getPracticeSession(user, sessionId);
  const questionIds = await ensurePracticeQuestionQueue(session);
  const answeredSummaryRows = await sql<Array<Pick<PracticeAnswer, 'question_id' | 'is_correct'>>>`
    SELECT question_id, is_correct
    FROM user_question_answers
    WHERE user_id = ${user.id}
      AND session_id = ${sessionId}
    ORDER BY answered_at, id
  `;
  const answeredSummary = new Map(answeredSummaryRows.map((answer) => [Number(answer.question_id), answer]));
  const requestedIndex = Number(params?.get('index'));
  const firstUnansweredIndex = questionIds.findIndex((questionId) => !answeredSummary.has(questionId));
  const fallbackIndex = firstUnansweredIndex >= 0 ? firstUnansweredIndex : 0;
  const currentIndex = Number.isInteger(requestedIndex)
    ? Math.max(0, Math.min(requestedIndex, Math.max(questionIds.length - 1, 0)))
    : fallbackIndex;
  const rows = await loadPracticeQuestionRows(session, questionIds, currentIndex, 1);
  const question = rows[0] ?? null;
  const questionId = question ? Number(question.question_id) : null;
  const result = questionId ? await getExistingPracticeAnswer(user.id, sessionId, questionId) as PracticeAnswer | null : null;
  const progress = buildPracticeProgress(questionIds, answeredSummary, currentIndex);

  return {
    session,
    question,
    questionIndex: question ? currentIndex : 0,
    total: questionIds.length,
    answeredCount: answeredSummary.size,
    progress,
    progressTruncated: progress.length < questionIds.length,
    result,
    previousIndex: currentIndex > 0 ? currentIndex - 1 : null,
    nextIndex: currentIndex < questionIds.length - 1 ? currentIndex + 1 : null
  };
}

function buildPracticeProgress(
  questionIds: number[],
  answered: Map<number, Pick<PracticeAnswer, 'question_id' | 'is_correct'>>,
  currentIndex: number
) {
  const total = questionIds.length;
  const indexes = total <= practiceProgressFullLimit
    ? questionIds.map((_, index) => index)
    : windowedProgressIndexes(total, currentIndex, practiceProgressWindowRadius);

  return indexes.map((index) => {
    const questionId = questionIds[index];
    const result = answered.get(Number(questionId));
    return {
      questionId,
      index,
      isAnswered: Boolean(result),
      isCorrect: result?.is_correct ?? null
    };
  });
}

function windowedProgressIndexes(total: number, currentIndex: number, radius: number) {
  const normalizedRadius = Math.max(1, Math.min(radius, Math.max(total - 1, 1)));
  const indexes = new Set<number>([0, total - 1, currentIndex]);
  const start = Math.max(0, currentIndex - normalizedRadius);
  const end = Math.min(total - 1, currentIndex + normalizedRadius);
  for (let index = start; index <= end; index += 1) {
    indexes.add(index);
  }
  return [...indexes].sort((left, right) => left - right);
}

export async function listPracticeSessions(user: User, params: URLSearchParams) {
  const limit = Math.min(Number(params.get('limit') || 20), 100);
  const status = params.get('status');
  return sql<PracticeSession[]>`
    SELECT *
    FROM user_practice_sessions
    WHERE user_id = ${user.id}
      AND (${status ?? null}::text IS NULL OR status = ${status ?? null})
    ORDER BY started_at DESC, id DESC
    LIMIT ${limit}
  `;
}

export async function getBankWrongQuestionCount(user: User, bankId: number) {
  await getBank(user, bankId);
  const rows = await sql<Array<{ count: number }>>`
    WITH active_question_ids AS (
      SELECT bql.question_id
      FROM bank_question_links bql
      JOIN questions q ON q.id = bql.question_id
      WHERE bql.bank_id = ${bankId}
        AND bql.status = 'active'
        AND q.status = 'active'

      UNION

      SELECT gql.question_id
      FROM bank_group_links bgl
      JOIN group_question_links gql ON gql.group_id = bgl.group_id
      JOIN questions q ON q.id = gql.question_id
      WHERE bgl.bank_id = ${bankId}
        AND bgl.status = 'active'
        AND q.status = 'active'
    )
    SELECT COUNT(DISTINCT ids.question_id)::int AS count
    FROM active_question_ids ids
    JOIN user_question_stats uqs
      ON uqs.question_id = ids.question_id
     AND uqs.user_id = ${user.id}
    WHERE uqs.wrong_count > 0 OR uqs.last_is_correct IS FALSE
  `;
  return rows[0]?.count ?? 0;
}

export async function submitAnswer(
  user: User,
  sessionId: number,
  data: { questionId: number; answerPayload: Record<string, unknown>; durationMs?: number }
) {
  const lock = await acquireRedisLock(redisKey('practice', sessionId, 'question', data.questionId, 'submit'), 10_000);
  if (!lock.acquired) {
    const existing = await getExistingPracticeAnswer(user.id, sessionId, data.questionId);
    if (existing) return existing;
    throw new ApiError(409, 'DUPLICATE_SUBMISSION', 'Answer submission is already in progress');
  }

  try {
    const existing = await getExistingPracticeAnswer(user.id, sessionId, data.questionId);
    if (existing) return existing;

    const session = await getPracticeSession(user, sessionId);
    if (session.status !== 'active') {
      throw new ApiError(409, 'INVALID_STATE', 'Practice session is not active');
    }
    const queuedQuestionIds = await redisGetJson<number[]>(practiceQuestionQueueKey(sessionId));
    if (queuedQuestionIds?.length && !queuedQuestionIds.map(Number).includes(Number(data.questionId))) {
      throw new ApiError(404, 'NOT_FOUND', 'Question is not part of this practice session');
    }

    const answerKey = await redisGetOrSetJson<{ id: number; answer_payload: string; score_payload: string } | null>(
      answerKeyCacheKey(data.questionId),
      Number(env.ANSWER_KEY_CACHE_TTL_SECONDS || 300),
      async () => {
        const answerKeyRows = await sql<Array<{ id: number; answer_payload: string; score_payload: string }>>`
          SELECT id, answer_payload, score_payload
          FROM question_answer_keys
          WHERE question_id = ${data.questionId}
            AND is_primary = true
          LIMIT 1
        `;
        return answerKeyRows[0] ?? null;
      }
    );
    const questionRows = await sql<Array<{ answer_mode: AnswerMode }>>`
      SELECT q.answer_mode
      FROM questions q
      WHERE q.id = ${data.questionId}
        AND q.status = 'active'
        AND EXISTS (
          SELECT 1
          FROM bank_question_links bql
          WHERE bql.bank_id = ${session.bank_id}
            AND bql.question_id = q.id
            AND bql.status = 'active'

          UNION ALL

          SELECT 1
          FROM bank_group_links bgl
          JOIN group_question_links gql ON gql.group_id = bgl.group_id
          WHERE bgl.bank_id = ${session.bank_id}
            AND gql.question_id = q.id
            AND bgl.status = 'active'
        )
      LIMIT 1
    `;
    if (!questionRows[0]) throw new ApiError(404, 'NOT_FOUND', 'Question is not part of this active practice session');
    validateSubmittedAnswerForTest(questionRows[0].answer_mode, data.answerPayload);

    const isCorrect = answerKey ? gradeAnswerForTest(questionRows[0].answer_mode, answerKey.answer_payload, data.answerPayload) : null;
    const maxScore = answerKey ? 1 : null;
    const score = isCorrect === null ? null : isCorrect ? 1 : 0;

    const rows = await sql`
      INSERT INTO user_question_answers (
        user_id, session_id, bank_id, question_id, answer_key_id,
        answer_payload, is_correct, score, max_score, duration_ms
      )
      VALUES (
        ${user.id}, ${sessionId}, ${session.bank_id}, ${data.questionId}, ${answerKey?.id ?? null},
        ${sql.json(toJsonValue(data.answerPayload))}, ${isCorrect}, ${score}, ${maxScore}, ${data.durationMs ?? null}
      )
      RETURNING *
    `;
    await invalidateUserAnalytics(user.id);
    await invalidateBankAnalytics(session.bank_id);
    await invalidateBankLeaderboard(session.bank_id);
    await invalidatePracticeSummaryCaches(user.id, session.bank_id);
    return rows[0];
  } finally {
    await lock.release();
  }
}

export function validateSubmittedAnswerForTest(mode: AnswerMode, actual: Record<string, unknown>) {
  if (mode === 'choice' && normalizeStringArray(objectValue(actual, 'selected')).length < 1) {
    throw new ApiError(422, 'VALIDATION_ERROR', 'Select at least one option before submitting');
  }
  if (mode === 'true_false' && typeof actual.value !== 'boolean') {
    throw new ApiError(422, 'VALIDATION_ERROR', 'Select true or false before submitting');
  }
  if (mode === 'fill_blank' && fillBlankValues(actual).length < 1) {
    throw new ApiError(422, 'VALIDATION_ERROR', 'Fill at least one blank before submitting');
  }
  if (mode === 'short_answer' && !hasNonEmptyText(actual.value)) {
    throw new ApiError(422, 'VALIDATION_ERROR', 'Answer text is required before submitting');
  }
}

export function gradeAnswerForTest(mode: AnswerMode, expectedJson: string, actual: Record<string, unknown>) {
  const expected = parseJson(expectedJson);
  if (!expected) return null;

  if (mode === 'choice') {
    return normalizeStringArray(objectValue(expected, 'selected')).join('|') === normalizeStringArray(objectValue(actual, 'selected')).join('|');
  }
  if (mode === 'true_false') {
    const expectedValue = objectValue(expected, 'value');
    const actualValue = objectValue(actual, 'value');
    if (typeof expectedValue !== 'boolean' || typeof actualValue !== 'boolean') return null;
    return expectedValue === actualValue;
  }
  if (mode === 'fill_blank') {
    const expectedValues = fillBlankValues(expected);
    const actualValues = fillBlankValues(actual);
    return expectedValues.length > 0
      && actualValues.length >= expectedValues.length
      && expectedValues.every((value, index) => value === actualValues[index]);
  }
  return null;
}

export async function completePracticeSession(user: User, sessionId: number, status: 'completed' | 'abandoned') {
  const rows = await sql<PracticeSession[]>`
    UPDATE user_practice_sessions
    SET status = ${status}, completed_at = NOW()
    WHERE id = ${sessionId}
      AND user_id = ${user.id}
      AND status = 'active'
    RETURNING *
  `;
  if (!rows[0]) throw new ApiError(404, 'NOT_FOUND', 'Active practice session not found');
  await invalidateUserAnalytics(user.id);
  return rows[0];
}

export async function getPracticeResults(user: User, sessionId: number) {
  await getPracticeSession(user, sessionId);
  return sql<Array<PracticeAnswer & { stem: string; answer_mode: AnswerMode; analysis: string | null; answer_keys: unknown }>>`
    SELECT
      uqa.*,
      q.stem,
      q.answer_mode,
      q.analysis,
      COALESCE(
        (
          SELECT json_agg(qak.* ORDER BY qak.version)
          FROM question_answer_keys qak
          WHERE qak.question_id = q.id
        ),
        '[]'
      ) AS answer_keys
    FROM user_question_answers uqa
    JOIN questions q ON q.id = uqa.question_id
    WHERE uqa.user_id = ${user.id}
      AND uqa.session_id = ${sessionId}
    ORDER BY uqa.answered_at, uqa.id
  `;
}

File: backend/src/practiq_backend/practice.py
from typing import Annotated, Literal

from fastapi import APIRouter, Query, Response
from psycopg.types.json import Jsonb
from pydantic import Field

from .content import details
from .core import DB, Error, Input, Page, Positive, bank, invalid, ok, one

router = APIRouter(prefix="/api/v1")


def grade(mode, expected, actual):
    if mode == "choice":
        return set(expected["correct"]) == set(actual["selected"])
    if mode == "true_false":
        return expected["answer"] == actual["value"]
    if mode == "fill_blank":

        def normalize(text):
            return " ".join(text.casefold().split())

        return [normalize(v) for v in expected["answers"]] == [normalize(v) for v in actual["value"]]
    return None


def session(db, sid, lock=False):
    s = one(
        db,
        "select s.* from practice_sessions s join question_banks b on b.id=s.bank_id where s.id=%s and b.deleted_at is null"
        + (" for update of s" if lock else ""),
        (sid,),
    )
    counts = one(
        db,
        "select count(*) question_count,(select count(*) from practice_answers where session_id=%s) answered_count from practice_session_questions where session_id=%s",
        (sid, sid),
    )
    scores = one(
        db,
        "select count(*) filter(where is_correct) correct_count,count(*) filter(where is_correct=false) wrong_count,coalesce(sum(score),0) score from practice_answers where session_id=%s",
        (sid,),
    )
    if hidden(s):
        scores = {key: None for key in scores}
    return {**s, **counts, **scores, "session_type": s["mode"]}


def hidden(s):
    return s["mode"] == "exam" and s["status"] != "completed"


def answer(db, s, qid):
    a = db.execute(
        "select a.*,k.answer_payload as answer_key_payload,k.explanation_payload from practice_answers a join question_answer_keys k on k.id=a.answer_key_id where session_id=%s and a.question_id=%s",
        (s["id"], qid),
    ).fetchone()
    if a and hidden(s):
        for field in (
            "answer_key_id",
            "is_correct",
            "score",
            "max_score",
            "answer_key_payload",
            "explanation_payload",
        ):
            a[field] = None
    return a


class Start(Input):
    bankId: Positive
    mode: Literal["all", "wrong", "by_type", "exam"] = "all"
    questionCount: Annotated[int, Field(ge=1, le=500)] = 30
    questionTypeId: str = ""
    allQuestions: bool = False


@router.post("/practice-sessions", status_code=201)
def start(body: Start, db: DB):
    bank(db, body.bankId)
    if body.mode == "by_type" and not body.questionTypeId:
        invalid("Choose a question type")
    rows = db.execute(
        """select q.id,k.id key_id from questions q join question_answer_keys k on k.question_id=q.id and k.is_primary
        where q.bank_id=%s and q.status='active' and q.deleted_at is null
        and (q.group_id is null or exists(select 1 from question_groups g where g.id=q.group_id and g.status='active'))
        and (%s<>'by_type' or q.question_type_id=%s)
        and (%s<>'wrong' or exists(select 1 from practice_answers a join practice_sessions s on s.id=a.session_id
             where a.question_id=q.id and a.is_correct=false and (s.mode<>'exam' or s.status='completed')))
        order by case when %s='exam' then random() else 0 end,q.sort_order,q.id limit %s""",
        (
            body.bankId,
            body.mode,
            body.questionTypeId,
            body.mode,
            body.mode,
            500 if body.allQuestions else body.questionCount,
        ),
    ).fetchall()
    if not rows:
        raise Error(409, "INVALID_STATE", "No active questions match this practice")
    s = one(
        db,
        "insert into practice_sessions(bank_id,mode) values(%s,%s) returning id",
        (body.bankId, "type" if body.mode == "by_type" else body.mode),
    )
    for i, q in enumerate(rows):
        db.execute(
            "insert into practice_session_questions(session_id,bank_id,question_id,position,answer_key_id) values(%s,%s,%s,%s,%s)",
            (s["id"], body.bankId, q["id"], i + 1, q["key_id"]),
        )
    return ok(session(db, s["id"]))


@router.get("/practice-sessions")
def sessions(db: DB, page: Page, status: Literal["active", "completed", "abandoned"] | None = None):
    ids = db.execute(
        "select s.id from practice_sessions s join question_banks b on b.id=s.bank_id where b.deleted_at is null and (%s::text is null or s.status=%s) order by s.id desc limit %s offset %s",
        (status, status, page.limit + 1, page.offset),
    ).fetchall()
    return page.result([session(db, i["id"]) for i in ids])


@router.get("/practice-sessions/{sid}")
def get_session(sid: Positive, db: DB):
    return ok(session(db, sid))


def practice_question(db, s, qid):
    q = details(db, qid)
    a = answer(db, s, qid)
    if a and not hidden(s):
        q["result"] = a
    return q


@router.get("/practice-sessions/{sid}/questions")
def session_questions(sid: Positive, db: DB):
    s = session(db, sid)
    return ok(
        [
            practice_question(db, s, r["question_id"])
            for r in db.execute(
                "select question_id from practice_session_questions where session_id=%s order by position",
                (sid,),
            )
        ]
    )


@router.get("/practice-sessions/{sid}/question-page")
def question_page(sid: Positive, db: DB, index: Annotated[int | None, Query(ge=0)] = None):
    s = session(db, sid)
    rows = db.execute(
        "select sq.question_id,a.id answer_id,a.is_correct from practice_session_questions sq left join practice_answers a on a.session_id=sq.session_id and a.question_id=sq.question_id where sq.session_id=%s order by sq.position",
        (sid,),
    ).fetchall()
    selected = (
        index if index is not None else next((i for i, q in enumerate(rows) if q["answer_id"] is None), 0)
    )
    if selected >= len(rows):
        invalid("Question index is out of range")
    qid = rows[selected]["question_id"]
    return ok(
        {
            "session": s,
            "question": practice_question(db, s, qid),
            "questionIndex": selected,
            "total": len(rows),
            "answeredCount": s["answered_count"],
            "result": answer(db, s, qid),
            "previousIndex": selected - 1 if selected else None,
            "nextIndex": selected + 1 if selected + 1 < len(rows) else None,
            "progress": [
                {
                    "index": i,
                    "questionId": q["question_id"],
                    "isAnswered": q["answer_id"] is not None,
                    "isCorrect": None if hidden(s) else q["is_correct"],
                }
                for i, q in enumerate(rows)
            ],
        }
    )


class Answer(Input):
    questionId: Positive
    answerPayload: dict
    durationMs: Annotated[int, Field(ge=0)] = 0


@router.post("/practice-sessions/{sid}/answers", status_code=201)
def submit(sid: Positive, body: Answer, db: DB):
    s = session(db, sid, True)
    old = answer(db, s, body.questionId)
    if old:
        return ok(old)
    if s["status"] != "active":
        raise Error(409, "INVALID_STATE", "Practice is no longer active")
    q = one(
        db,
        "select q.*,k.id key_id,k.answer_payload expected from practice_session_questions sq join questions q on q.id=sq.question_id join question_answer_keys k on k.id=sq.answer_key_id where sq.session_id=%s and sq.question_id=%s",
        (sid, body.questionId),
    )
    payload = body.answerPayload
    mode = q["answer_mode"]
    if mode == "choice":
        selected = payload.get("selected")
        labels = {
            o["option_label"]
            for o in db.execute(
                "select option_label from question_options where question_id=%s", (body.questionId,)
            )
        }
        if (
            not isinstance(selected, list)
            or not selected
            or any(not isinstance(v, str) for v in selected)
            or not set(selected) <= labels
            or len(set(selected)) != len(selected)
            or (q["choice_variant"] == "single" and len(selected) != 1)
        ):
            invalid("Choose valid options")
    elif mode == "true_false":
        if not isinstance(payload.get("value"), bool):
            invalid("Choose true or false")
    elif mode == "fill_blank":
        v = payload.get("value")
        if not isinstance(v, list) or not v or any(not isinstance(x, str) or not x.strip() for x in v):
            invalid("Fill the blanks")
    elif not isinstance(payload.get("value"), str) or not payload["value"].strip():
        invalid("Enter an answer")
    result = grade(mode, q["expected"], payload)
    db.execute(
        "insert into practice_answers(session_id,question_id,answer_key_id,answer_payload,is_correct,score,duration_ms) values(%s,%s,%s,%s,%s,%s,%s)",
        (
            sid,
            body.questionId,
            q["key_id"],
            Jsonb(payload),
            result,
            None if result is None else int(result),
            body.durationMs,
        ),
    )
    db.execute("update practice_sessions set updated_at=now() where id=%s", (sid,))
    return ok(answer(db, s, body.questionId))


@router.post("/practice-sessions/{sid}/{action}")
def finish(sid: Positive, action: Literal["complete", "abandon"], db: DB):
    s = session(db, sid, True)
    target = "completed" if action == "complete" else "abandoned"
    if s["status"] != "active":
        raise Error(409, "INVALID_STATE", "Practice is no longer active")
    db.execute(
        "update practice_sessions set status=%s,completed_at=now(),updated_at=now() where id=%s",
        (target, sid),
    )
    return ok(session(db, sid))


@router.get("/practice-sessions/{sid}/results")
def results(sid: Positive, db: DB):
    s = session(db, sid)
    if s["status"] == "active":
        raise Error(409, "INVALID_STATE", "Complete the session before viewing results")
    return ok(
        [
            {**practice_question(db, s, r["question_id"]), "result": answer(db, s, r["question_id"])}
            for r in db.execute(
                "select question_id from practice_session_questions where session_id=%s order by position",
                (sid,),
            )
        ]
    )


@router.delete("/banks/{bank_id}/practice-data", status_code=204)
def reset(bank_id: Positive, db: DB):
    bank(db, bank_id)
    db.execute("delete from practice_sessions where bank_id=%s", (bank_id,))
    return Response(status_code=204)


VISIBLE_ANSWERS = """from practice_answers a join practice_sessions s on s.id=a.session_id
    join question_banks b on b.id=s.bank_id join questions q on q.id=a.question_id
    where b.deleted_at is null and q.deleted_at is null and (s.mode<>'exam' or s.status='completed')"""


def summary(db, bank_id=None):
    stats = one(
        db,
        "select count(*) attempts,count(*) filter(where a.is_correct) correct,count(*) filter(where a.is_correct=false) wrong "
        + VISIBLE_ANSWERS
        + " and (%s::bigint is null or s.bank_id=%s)",
        (bank_id, bank_id),
    )
    stats["accuracy"] = round(100 * stats["correct"] / stats["attempts"], 1) if stats["attempts"] else 0
    stats.update(
        one(
            db,
            "select count(*) banks,count(*) filter(where is_favorite) favorite_banks from question_banks where deleted_at is null",
        )
    )
    return stats


@router.get("/analytics/summary")
def get_summary(db: DB):
    return ok(summary(db))


@router.get("/analytics/snapshot")
def snapshot(db: DB):
    ids = db.execute(
        "select s.id from practice_sessions s join question_banks b on b.id=s.bank_id where b.deleted_at is null order by s.id desc limit 10"
    ).fetchall()
    weak = db.execute(
        "select q.id question_id,q.stem,s.bank_id,count(*) attempt_count,count(*) filter(where a.is_correct=false) wrong_count,avg(case when a.is_correct then 1.0 else 0 end) mastery_score "
        + VISIBLE_ANSWERS
        + " group by q.id,s.bank_id order by mastery_score,wrong_count desc limit 10"
    ).fetchall()
    trend = db.execute(
        "select a.answered_at::date as day,count(*) attempts,count(*) filter(where a.is_correct) correct "
        + VISIBLE_ANSWERS
        + " group by day order by day desc limit 30"
    ).fetchall()
    return ok(
        {
            "summary": summary(db),
            "recentSessions": [session(db, r["id"]) for r in ids],
            "weakQuestions": weak,
            "trend": trend,
        }
    )


@router.get("/analytics/banks/{bank_id}")
def bank_analytics(bank_id: Positive, db: DB):
    bank(db, bank_id)
    return ok(summary(db, bank_id))
